### PR TITLE
[JUJU-824] Log output via syslog

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following allows logging from the logsink to point to a local syslog.
This change aims to help with large deployments that don't want to use
the database as a store for logs. Instead all the logs will be written to
the controllers syslog under the `juju,daemon` tag.

The aim of this is to reduce the work that the database requires for
managing juju.

One important fact here, you can **no longer** use `juju debug-log` if
you go all-in with syslog for that model.

If you want to test it, you can write to both the database and syslog at
the same time and then turn off either if required.

`$ juju model-config -m controller logging-config="database,syslog"`

## QA steps

```sh
$ juju bootstrap lxd test --config="logging-config=syslog"
$ lxc exec <container id> -- bash
$ journalctl -f
```

Switching the logging-output of an existing deployment will require
the restarting of all controller agents.

```sh
$ juju bootstrap lxd test
$ juju model-config -m controller logging-config="syslog"
$ lxc exec <container id> -- bash
$ sudo systemctl restart jujud-*
$ journalctl -f
```

### CODE REVIEW

File: environs/config/config.go [hunk 1]
```diff
@@ -2234,4 +2279,9 @@ where possible. (default "")`,
 		Type:        environschema.Tstring,
 		Group:       environschema.EnvironGroup,
 	},
+	LoggingOutputKey: {
+		Description: `The logging output destination: database and/or syslog. (default "")`,
```
Comment (hmlanigan) [inline environs/config/config.go:2283]: Calling the traditional logging type database, feels like an implementation detail that user's wouldn't need to know and potentially confusing to new users. Pondering alternatives.

**Review by hmlanigan — COMMENTED** [review]
Most QA good

File: environs/config/config.go (hunk 1 above)
Comment (SimonRichardson) [reply environs/config/config.go:2283]: My original thought was `state`

File: core/logger/loggers.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+// MakeLoggers creates loggers from a given LoggersConfig.
+func MakeLoggers(outputs []string, config LoggersConfig) LoggerCloser {
+	results := make(map[string]Logger)
+loop:
+	for _, output := range outputs {
+		switch output {
+		case SyslogName:
+			results[SyslogName] = config.SysLogger()
+		default:
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply core/logger/loggers.go:34]: Ah, yes I'll fix that

File: environs/config/config.go [hunk 3]
```diff
@@ -1410,6 +1420,40 @@ func (c *Config) validateMode() error {
 	return nil
 }
 
+// LoggingOutput is a for determining the destination of output for
+// logging.
+func (c *Config) LoggingOutput() ([]string, bool) {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply environs/config/config.go:1425]: It's convention in here, even if it's not actually used outside of tests. Might be worth having another look at this in the future

**Comment by SimonRichardson** [comment]
@hmlanigan I think we should go with database as nobody has come forward to replace it.

**Review by hmlanigan — APPROVED** [review]
We've had no additional ideas of a replacement for `database`, so let's move forward.

File: core/logger/tee.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	"io"
+)
+
+// TeeLogger forwards log request to each underlying logger.
+type TeeLogger struct {
+	loggers []Logger
+}
+
+// NewTeeLogger returns a logger that forwards log requests to each one of the
+// provider loggers.
```
Comment (manadart) [inline core/logger/tee.go:16]: "provided"?

File: worker/syslogger/worker.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func (w *syslogWorker) Log(logs []corelogger.LogRecord) error {
+	for _, log := range logs {
+		writer, ok := w.writers[log.Level]
+		if !ok {
+			continue
+		}
+		dateTime := log.Time.In(time.UTC).Format("2006-01-02 15:04:05")
+		fmt.Fprintf(writer, "%s %s %s %s\n", dateTime, log.Entity, log.Module, log.Message)
```
Comment (manadart) [inline worker/syslogger/worker.go:127]: Unchecked error. 

I think you should have a test that verifies satisfactory behaviour for killing the worker while it's writing logs.

**Review by manadart — APPROVED** [review]
Approved with a caveat regarding the syslogger worker.

File: apiserver/logsink.go [hunk 6]
```diff
@@ -17,64 +17,62 @@ import (
...
+	loggingOutputs      []string
+	clock               clock.Clock
+	loggerBufferSize    int
+	loggerFlushInterval time.Duration
+	mu                  sync.Mutex
+	loggers             map[*state.State]*bufferedLogger
 }
 
-func (d *dbloggers) get(st *state.State) recordLogger {
+func (d *apiServerLoggers) get(st *state.State) RecordLogger {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/logsink.go]: getLogger as it's not actually a DB logger exclusively.

File: apiserver/logsink.go [hunk 7]
```diff
@@ -17,64 +17,62 @@ import (
...
-	loggers               map[*state.State]*bufferedDbLogger
+// state pool, the strategies must call the apiServerLoggers.remove method.
+type apiServerLoggers struct {
+	syslogger           syslogger.SysLogger
+	loggingOutputs      []string
+	clock               clock.Clock
+	loggerBufferSize    int
+	loggerFlushInterval time.Duration
+	mu                  sync.Mutex
+	loggers             map[*state.State]*bufferedLogger
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/logsink.go:50]: I wonder if `st.ModelUUID()` would actually be better here.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**